### PR TITLE
Khalyla requirement

### PR DESCRIPTION
**Pull request by petemeehan**
requirement:
1. given a sample set of pledges, choose the winning applicants. the total must be £2.5m or as close to that as possible, but not more
2. handle the case where some applicants are unable to pay by choosing new applicants from the original set. also this could happen multiple times in the future

### CODE REVIEW

File: monzo.js [hunk 1]
```diff
@@ -1,20 +1,83 @@
 var PD = require("probability-distributions");
 
-var goal = 25000;
-var pledgeValues = [10,20,50,100,250,500,1000];
+var goal = 2500;
```
Comment (cmck) [inline monzo.js]: please change to 2,500,000 and retest

File: monzo.js [hunk 2]
```diff
@@ -1,20 +1,83 @@
...
+var pledgeValues = [10,20,50,100];
```
Comment (cmck) [inline monzo.js]: this has changed, why? people can pledge 10,20,50,100,250 or 1000

File: monzo.js [hunk 3]
```diff
@@ -1,20 +1,83 @@
...
 var datasetTotal = dataset.reduce((a, b) => a + b, 0);
-var randomlyChosenPledge;
-while (datasetTotal >= goal ) {
-	randomlyChosenPledge = dataset.splice(
-    Math.floor(Math.random() * dataset.length), 1
-  )[0];
-	datasetTotal -= randomlyChosenPledge;
+
+/*  Convert array to object so each pledge has an ID */
+dataset = dataset.reduce(function(obj, currentValue, index) {
```
Comment (cmck) [inline monzo.js]: it's not clear that this is test data

File: monzo.js [hunk 4]
```diff
@@ -1,20 +1,83 @@
 var PD = require("probability-distributions");
 
-var goal = 25000;
+var goal = 2500000;
 var pledgeValues = [10,20,50,100,250,500,1000];
 var maxPledgeValue = pledgeValues[pledgeValues.length - 1];
-var dataset = PD.sample(pledgeValues, 1000, true, [7,6,5,4,3,2,1]);
+var dataset = PD.sample(pledgeValues, 100000, true, [7,6,5,4,3,2,1]);
```
Comment (cmck) [inline monzo.js]: this line fails with a sample set size of 1000. The code should work regardless of the number of applicants

File: monzo.js [hunk 5]
```diff
@@ -1,20 +1,83 @@
...
 
-var goal = 25000;
-var pledgeValues = [10,20,50,100,250,500,1000];
+var goal = 2500;
+var pledgeValues = [10,20,50,100];
 var maxPledgeValue = pledgeValues[pledgeValues.length - 1];
-var dataset = PD.sample(pledgeValues, 1000, true, [7,6,5,4,3,2,1]);
+var dataset = PD.sample(pledgeValues, 1000, true, [7,6,5,4]);
 
 var datasetTotal = dataset.reduce((a, b) => a + b, 0);
```
Comment (cmck) [inline monzo.js]: ES6 only, you need a solution to this if the script is going to be used in a browser that doesn't support it

File: monzo.js [hunk 6]
```diff
@@ -1,20 +1,83 @@
...
+
+/*  Convert array to object so each pledge has an ID */
+dataset = dataset.reduce(function(obj, currentValue, index) {
+    obj[index] = currentValue;
+    return obj;
+}, {});
+
+/*  Make copy of original dataset and total and randomly select pledges to
+    to remove so it now contains only chosen pledges */
+var chosenDatasetTotal = datasetTotal;
```
Comment (cmck) [inline monzo.js]: move algorithm to its own function

File: monzo.js [hunk 7]
```diff
@@ -1,20 +1,83 @@
...
+var chosenDataset = {};
+for (var i in dataset) chosenDataset[i] = dataset[i]; // replace with Object.assign()?
+while (chosenDatasetTotal >= goal ) {
```
Comment (cmck) [inline monzo.js]: bug when goal=2.5m, dataset length 10k. you need to handle the case when less than 2.5m is raised

File: monzo.js [hunk 8]
```diff
@@ -1,20 +1,83 @@
...
+  var ids = Object.keys(chosenDataset);
```
Comment (cmck) [inline monzo.js]: indenting

File: monzo.js [hunk 9]
```diff
@@ -1,20 +1,83 @@
...
+	var randomId = ids[Math.floor(Math.random() * ids.length)];
+	chosenDatasetTotal -= chosenDataset[randomId];
+  delete chosenDataset[randomId];
 }
+/*  Update original dataset by removing chosen applicants, so they can't
+    be chosen again later if more applicants are required */
+for (var i in chosenDataset) delete dataset[i];
```
Comment (cmck) [inline monzo.js]: move into its own function

File: monzo.js [hunk 10]
```diff
@@ -1,20 +1,83 @@
...
 
-console.log("Goal: " + goal);
+console.log("\nGoal: " + goal);
 console.log("Total pledged: " + datasetTotal);
```
Comment (cmck) [inline monzo.js]: look into the comma separation for string formatting in console.log()

File: monzo.js [hunk 11]
```diff
@@ -1,20 +1,83 @@
...
-console.log("Number of chosen pledgers: " + dataset.length);
-console.log("Chosen pledges total: " + datasetTotal);
+console.log("Number of chosen pledgers: " + Object.keys(chosenDataset).length);
+console.log("Chosen pledges total: " + chosenDatasetTotal);
+
+
+
+/*  Remove failed pledges */
+var numberOfDroppedPledgers = 10;
```
Comment (cmck) [inline monzo.js]: numberOfDroppedPledgers needs to be 10% of the set of applications

File: monzo.js [hunk 12]
```diff
@@ -1,20 +1,83 @@
...
+/*  Repopulate chosen pledges dataset */
+repopulateChosenDataset();
+
+console.log("\nNEW PLEDGERS SELECTED");
+console.log("New number of chosen pledgers: " + Object.keys(chosenDataset).length);
+console.log("New chosen pledges total: " + chosenDatasetTotal);
+
+
+
+function getRandomSubsetOfChosenApplicants(obj, numberOf) {
```
Comment (cmck) [inline monzo.js]: unclear what obj parameter is

File: monzo.js (hunk 7 above)
Comment (cmck) [reply monzo.js]: also this section of code is slow try to work out why

File: monzo.js [hunk 13]
```diff
@@ -1,20 +1,83 @@
...
+  var keys = Object.keys(obj);
+  for (var i = 0; i < numberOf; i++) {
+    var randomIndex = Math.floor(Math.random() * keys.length);
+    var randomKey = keys.splice(randomIndex, 1)[0];
+    subset.push(randomKey);
+  }
+  return subset;
+}
+
+function removeSubsetOfApplicants(obj, arrayOfApplicantIds) {
```
Comment (cmck) [inline monzo.js]: document these functions

**Comment by cmck** [comment]
here's a failure case hope it helps, 

```
var goal = 2500000;
var pledgeValues = [10,20,50,100,250,500,1000];
var maxPledgeValue = pledgeValues[pledgeValues.length - 1];
var dataset = PD.sample(pledgeValues, 30000, true, [7,6,5,4,3,2,1]);
```

```
C:monzo-crowdfund clark$ node monzo.js
main algorithm Took **21635.5000 milliseconds**

Goal: 2500000
Total pledged: 3804720
Number of chosen pledgers: 19442
Chosen pledges total: 2499860
getRandomSubsetOfChosenApplicants Took 1.5850 milliseconds
removeSubsetOfApplicants Took 0.1320 milliseconds

OH FUCK, WE LOST 10 PLEDGERS!
New number of chosen pledgers: 19432
New chosen pledges total: 2499370
repopulateChosenDataset Took 0.0820 milliseconds

NEW PLEDGERS SELECTED
New number of chosen pledgers: 19432
New chosen pledges total: 2499370
```

File: monzo.js (hunk 8 above)
Comment (petemeehan) [reply monzo.js]: elaborate?

File: monzo.js (hunk 4 above)
Comment (petemeehan) [reply monzo.js]: it fails when not enough initial pledges to reach goal. this is a known bug. fix coming soon

File: monzo.js (hunk 8 above)
Comment (cmck) [reply monzo.js]: `var randomId = ids[Math.floor(Math.random() * ids.length)]` shouldn't be indented.

File: monzo.js (hunk 13 above)
Comment (petemeehan) [reply monzo.js]: please advise whether new comments are sufficient

File: monzo.js (hunk 8 above)
Comment (petemeehan) [reply monzo.js]: apologies for not noticing this. fixed

**Review by petemeehan — COMMENTED** [review]
Need to catch exception if goal not reachable

**Comment by cmck** [comment]
Lgtm

**Pull request merged**